### PR TITLE
Fix batching for nested flatmaps at different levels of composition

### PR DESCRIPTION
**Pull request by stevenheidel**
A coworker found an edge case that meant we weren't optimally batching in the flushDownstream method. I had to make the `getClumpsByLevel` a bit more sophisticated to handle this.

### CODE REVIEW

File: src/test/scala/io/getclump/ClumpExecutionSpec.scala [hunk 1]
```diff
@@ -48,9 +48,22 @@ class ClumpExecutionSpec extends Spec {
...
+
+      clumpResult(Clump.collect(clump1, clump2)) mustEqual Some(List(100, 200))
+      source1Fetches mustEqual List(Set(1, 2))
+      source2Fetches mustEqual List(Set(20, 10))
+    }
+
+    // Implementation note: this test will fail if ClumpContext::getClumpsByLevel does not satisfy the requirement that
+    // "Clumps appear as early in the list as possible"
+    "for clumps created inside nested flatmaps at different levels of composition" in new Context {
+      val clump1 = Clump.value(1).flatMap(source1.get).flatMap(source2.get).map(identity)
```
Comment (stevenheidel) [inline src/test/scala/io/getclump/ClumpExecutionSpec.scala]: Adding the `.map(identity)` here used to make it quit correctly batching requests after the first one. It would still get the right answer but it was doing too many calls to the source.

**Comment by stevenheidel** [comment]
@fwbrasil @williamboxhall - look good to you?

**Comment by fwbrasil** [comment]
:+1:
